Feature/enable skbuild

vyasr opened the pull request:
<!-- Provide a general summary of your changes in the Title above. -->

## Description
<!-- Describe your changes in detail. -->

Use scikit-build and CMake to build the C++ and Cython components of freud.

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- Replace ??? with the issue number that this pull request resolves, if applicable. -->

This pull request dramatically speeds up builds of freud (takes < 30 seconds on my Mac using Ninja). Changing to a proper build system should also dramatically simplify addressing issues like #464 and #629. It will also enable easier integration with IDEs for the C++ code, and long term it will enable exposing a C++ API for freud.

Outstanding tasks:

- [x] Enable automatic download of submodules.
- [x] Contact voro++ maintainer to discuss modifications that would allow using it as a library (rather than having to handle compiling its sources ourselves).
- [x] Update CI scripts to work with new build system.
- [x] Update and test deployment (to Test PyPI).
- [x] Update development requirements.
- [x] Update documentation on building freud.


**Update** -- The comments below are mostly outdated now due to later changes to the CMake configuration, but I'll leave them here so that we can track the discussions in the future if needed.

This PR supersedes #661. I glanced through that PR, and I note that this implementation is different: rather than compiling a single shared object for the entire C++ library, I am instead following a more distributed model where each Cython module is compiled against the corresponding C++ code. We can change this in the future if we move towards a more thorough separation of the C++ code and exposing those APIs, but for now I'd prefer to retain this model since it's cleaner and generates leaner builds.

@joaander as a more experienced CMake user, some specific questions for you (in addition to any review you can provide).
- What is a reasonable minimum version of CMake to use? I've looked through the changelog etc briefly, but I'd prefer a more informed opinion.
- If you look at `freud/CMakelists.txt`, I'm forced to link the `_util` Object library directly to a couple of Cython modules rather than to the corresponding C++ Object libraries (i.e. `_environment` instead of `environment`). I _think_ that the reason is that the relevant C++ util code (the `diagonalize.cc` source file) is only included by the source files of the corresponding C++ modules, not the headers. As a result, I've marked the corresponding target includes (for instance, in `cpp/environment/CMakeLists.txt`) as PRIVATE; however, I note that changing this to PUBLIC doesn't fix the issue, so I'm not certain this is correct. It may also be some limitation with Object library includes propagation that I'm not finding in the documentation. Any thoughts?
- I wrote a simple FindTBB.cmake script, but I know that we discussed the more sophisticated version along with the config file that you have on HOOMD's next branch. Would you recommend I try to copy in that file and the corresponding macro?

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to
     see how your changes affect other areas of the code, etc. -->

Existing builds on my machine are functional. Additional testing will come in the form of the to-do list above.

**Update** - Building and testing on CI now works on all systems. The wheel building process has also been manually verified via push to test PyPI.

## Screenshots
<!-- (if appropriate) -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
<!-- Put an `x` in all the boxes that apply. If you're unsure about any of
     these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

vyasr committed:
Ignore skbuild files.

vyasr committed:
Create CMakeLists for locality cpp.

vyasr committed:
Add overall cpp CMakeLists.

vyasr committed:
Add overall CMakeLists.

vyasr committed:
Move includes to top level for use with Cython builds.

vyasr committed:
Add CMakeLists for freud with just locality for now.

vyasr committed:
Add FindTBB to actually load the library, shuffle around and cleanup imports to get successful build.

vyasr committed:
Get voro++ to build so we can link to it (currently requires a minor hack to voro++ CMakeLists, will need to resolve eventually).

vyasr committed:
Make the build link properly against voro++.

vyasr committed:
Add box build.

vyasr committed:
Enable cluster module.

vyasr committed:
Do some cleanup looping over modules to facilitate adding more.

vyasr committed:
Enable density module.

vyasr committed:
Build diagonalize in util.

vyasr committed:
Enable environment module.

vyasr committed:
Enable pmft.

vyasr committed:
Enable order.

vyasr committed:
Enable parallel.

vyasr committed:
Create util extension module.

vyasr committed:
Integrate non-CPP modules into primary install loop.

vyasr committed:
Enable all remaining non-CPP modules.

vyasr committed:
Reenable Cython compile flags.

vyasr committed:
Add back compiler macros.

vyasr committed:
Silence deprecation warnings arising from Python3.8/Cython incompatibility issues.

vyasr committed:
Add back Windows options.

vyasr committed:
Add option for coverage and line tracing.

vyasr committed:
Create subdirectory for CMake scripts.

vyasr committed:
Clean up TBB usage.

vyasr committed:
Use public links for TBB and add notes regarding how links to internal object libraries aren't working.

vyasr committed:
Clarify usage of PRIVATE libraries for _util (which is not imported in headers).

vyasr committed:
Minimize PUBLIC targets.

vyasr committed:
Switch to directly compiling required voro++ source.

vyasr committed:
Minor reorganization and add option for build type.

vyasr committed:
Update pyproject.toml with new build requirements.

vyasr committed:
Only include extern where necessary.

bdice commented on CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+cmake_minimum_required(VERSION 3.10)
```
I've seen recommendations to use at least 3.14 for "modern" features. I used 3.16 in #661.

vyasr replied:
Right, on the flip side I don't know how onerous that is. 3.14 did only come out last year, I assume that for instance Ubuntu or Arch package managers for anything other than the newest OS versions might not have a sufficiently new CMake. Just want to make sure we're not causing problems for installation on e.g. any of our clusters like Comet.

vyasr replied:
To be clear: you are absolutely right that I am using features that require 3.14 (all the `target_includes` that I'm doing don't exist in 3.10 AFAIK). I just want to make sure that won't be a problem for our HPC user base. HOOMD is still requiring only 3.9, although I don't know if that's just a legacy thing or planned to change for the v3 release.

joaander replied:
HOOMD v2.x builds with CMake 2.8.10.1 or newer. HOOMD v3.x will bump that requirement to 3.9. 3.9 was the first version to have a functioning findMPI that uses modern practices. I suggest supporting the oldest version you can without needing to implement excessive workarounds. This makes the software accessible to more users, many of whom do not update their software stacks that regularly.

bdice replied:
CMake version pi.0 (3.14.0) was [announced on Pi Day 2019](https://blog.kitware.com/cmake-3-14-0-available-for-download/). I'm fine with matching HOOMD's requirement of CMake 3.9 as long as it doesn't require significant workarounds. We provide conda-forge builds and/or wheels for freud on multiple platforms and architectures, so far fewer users will build from source compared to HOOMD. Also since CMake is installable with pip or conda, I think it would be acceptable to have a version requirement 3.14 that is now >18 months old.

vyasr replied:
With minor modifications we're able to support CMake >= 3.3. I modified CI to test against 3.6.3, which is the oldest version available on PyPI.

bdice commented on cpp/environment/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+    LocalDescriptors.h
+    LocalDescriptors.cc
+    MatchEnv.h
+    MatchEnv.cc
```
Missing `Registration.h`? Or is that not necessary?
```suggestion
    MatchEnv.cc
    Registration.h
```

vyasr replied:
Good catch. It's not strictly necessary since it's header only and the whole cpp/environment is included, but better to be explicit here.

bdice commented on cpp/locality/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+    Voronoi.cc
+    Voronoi.h
+
+    # For now, compile voro++ object in directly.
```
There are CMake configuration files for voro++. Do they need to be edited, or just called correctly?

vyasr replied:
They need to be edited. The current voro++ CMake files are not configured properly for building as a library in another code base, see the to-do list in the PR description.

bdice commented on setup.py:
```diff
@@ -419,100 +13,53 @@ def find_tbb(tbb_root=None, tbb_include=None, tbb_link=None):
+        'Issue Tracker': 'https://github.com/glotzerlab/freud/issues',
+    },
+    python_requires='>=3.6',
+    install_requires=[
```
Does this need to list scikit-build and cmake? Not sure how these interact with pyproject.toml.

vyasr replied:
No, the `install_requires` is to determine dependencies of the code, whereas pyproject.toml is for dependencies of the build process. The requirements in pyproject.toml indicate what needs to be installed to even run setup.py.

bdice reviewed (commented):
Cool, @vyasr. This will help the library a lot. I have a few questions you might be able to answer.

bdice commented:
@vyasr Also I didn't understand your reasoning for the distributed approach: "it's cleaner and generates leaner builds" -- how so? I would have thought the other approach which compiles one shared object for the whole C++ library and then just links it to the Cython modules would have been cleaner/leaner. Happy to discuss on Slack/call if it's a long explanation.

vyasr commented:
Regarding the leaner/cleaner argument, I should clarify. My primary objection is to having one monolithic CMakeLists file. That's poor CMake style, harder to read, and more difficult to integrate into complex builds. It also makes it less likely that you'll properly isolate build requirements for different pieces of the library. It's much easier to maintain if we have separate CMakeLists for each module.

As far as the result of separate libraries being leaner, it's primarily a product of the fact that building the underscore Cython modules requires pulling a lot of the cpp header files and results in duplication. Also, it adds an extra file that needs to be installed with the build (the libfreud.so library). My usage of object libraries in CMake means that the only actual library files created are the ones required for Cython. However, I don't think either any of this really matters: the difference in total filesize is very small, the compile times will be almost identical (all the compilation is identical), and the extra file is necessary anyway if we ever want to expose a libfreud module. That being the case, I can go ahead update `cpp/CMakeLists` to build a freud C++ library now. The object file methods make it easy to mostly get the best of both worlds (separate CMakeLists files while generating a single large library).

bdice commented:
@vyasr That makes a lot more sense. I support your argument for many CMake list files. I do think separating “libfreud” as a C++ library would be a positive change. That should also simplify TBB linkage, I think? (The Cython modules should just link to libfreud, which links to TBB.)

vyasr committed:
Change to building a single shared library for C++ code.

vyasr committed:
Add missing Registration header.

vyasr commented:
No, unless you mean in the context of the CMake configuration it doesn't really change linking. You'll still need to link all the Cython libraries against TBB as well, shared library linking isn't transitive in that way. CMake handles this appropriately though.

vyasr committed:
Automate submodule initialization. Resolves #571.

vyasr committed:
Add new build requirements to requirements-testing.

vyasr committed:
Bump up cmake wheel version.

vyasr committed:
Make changes necessary for coverage to actually run, and modify CI config appropriately.

joaander commented on CMake/UpdateSubmodules.cmake:
```diff
@@ -0,0 +1,14 @@
+	option(UPDATE_SUBMODULES "Download missing submodules during build" ON)
+	if (UPDATE_SUBMODULES )
+		message("Updating submodules.")
+		execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive
```
I removed this from HOOMD v3. I found it caused more problems than it solved. Users and developers should manage their git repositories themselves.

vyasr replied:
Yes I recall you doing this. I'll probably replace this with an immediate error so that it's more obvious to the user why it's failing (rather than throwing a compile error that a file is missing). Could you elaborate on what issues you ran into with this approach?

bdice replied:
I agree with @joaander - it gets clunky when changing submodules.

joaander replied:
As a developer, it is frustrating when you make a change to submodule, compile, and then that change is gone. It is also frustrating that the very slow `git submodule update --init` process occurs every time you compile. As a user who builds the code only once, it is a minor teaching moment to learn to use `git clone --recursive` or `git submodule update --init`.

vyasr replied:
Makes sense. The current build script just raises errors to indicate when submodules are not present, which I think is a useful compromise.

joaander commented on CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+endif()
+
+set(CMAKE_CXX_STANDARD 11)
```
Can you bump this to 14? TBB will be requiring C++14 soon.

vyasr replied:
I don't see any reason that we can't. @bdice I assume you would be happy with these changes? That would also allow us to use at least some of the newer algorithms in the standard library :)

bdice replied:
```suggestion
set(CMAKE_CXX_STANDARD 14)
```

joaander commented on CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
```
This shouldn't be necessary. CMake should use the right compile flags depending on the type of library being built.

vyasr replied:
Got it, I had issues with it early on but it might have been because I was misusing OBJECT and SHARED in my early iterations. It does indeed seem to work correctly on shared libraries without this.

joaander commented on cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+# Set the name of the library to just freud.
+set_target_properties(
+    libfreud
+    PROPERTIES OUTPUT_NAME freud
```
Why rename this. It doesn't appear to be an importable python module.

vyasr replied:
It's not, however because it's a shared library and not a Python extension module CMake automatically prepends `lib` to the name. A `liblibfreud.so` seems very strange, and will also not be what people expect if we get around to exposing our C++ API. I don't want to name the actual library (in CMake) freud, though, because that conflicts with the project name and is confusing (not to mention I'm not 100% sure that CMake will always disambiguate correctly).

bdice replied:
@vyasr I would recommend clarifying this in the comment. I found that out earlier while playing with CMake and forgot that would happen. Maybe update the comment to say:
```
# CMake prepends 'lib' to the shared library name.
# This sets the name of the library to libfreud instead of liblibfreud.
```

vyasr replied:
Good idea, I added more explicit documentation of this.

joaander replied:
Most old-style CMake scripts name the package in all caps to avoid the name conflict. Most new-style scripts use namespaces. You can name the target `freud::freud` and then the library should be named `libfreud` without needing `set_target_properties`

vyasr replied:
It looks like that namespacing is not supported for targets that must be built, only for imported ones: https://gitlab.kitware.com/cmake/cmake/-/issues/20444. So it works for TBB, but not for freud.

joaander reviewed (commented):
Looking good. I left a few comments and questions.

Regarding `_util` I'm not sure what the problem is. As you say, maybe it is some limitation with object libraries. Since you are building everything into one `libfreud` target, you could instead make the child CMakeList.txt files set a list of source files in a variable and then have the `add_library` compile all those source files together. That would work around any limitations with `OBJECT` and wouldn't result in any files being compiled more than once.

Regarding FindTBB, feel free to use the solution from HOOMD to support finding modern TBB's by their embedded config file (the `find_package_config_first` macro) - that way you don't need to rely on your own `FindTBB.cmake` except on systems like Ubuntu which doesn't provide the TBB config or on platforms with older TBB installations. The `FindTBB` from HOOMD determines the `TBB_VERSION` and implements a few compiler workarounds that may not be needed for freud. The only other main difference is that the target is `TBB::tbb` for compatibility with the TBB provided config files.

vyasr committed:
Fix variable setting.

vyasr commented:
@jglaser might also be interested in this.

vyasr committed:
Stop automatically cloning submodules, just raise an error.

vyasr committed:
Ignore cxx as well as cpp files.

vyasr committed:
Remove extraneous fPIC switch.

vyasr committed:
Switch to C++14.

vyasr committed:
Some minor changes to support CMake >= 3.3.0.

vyasr committed:
Update testing requirements with minimum cmake version testable.

vyasr committed:
Reduce number of threads on CI.

vyasr committed:
Use newer cmake.

vyasr committed:
Request verbose output.

vyasr committed:
Require fPIC.

bdice commented on .gitignore:
```diff
@@ -14,6 +14,9 @@ doc/build/*
 freud/*.so
 freud/*.cpp
+freud/*.cxx
+freud/libfreud.dylib
```
Would prefer to match the catch-all style. We shouldn't ever commit a dylib file.
```suggestion
freud/*.dylib
```

bdice commented on CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+    voro++)
+foreach (submodule ${submodules})
+    if( NOT EXISTS "${PROJECT_SOURCE_DIR}/extern/${submodule}/.git" )
+        message( FATAL_ERROR "${Red}The git submodules are not available. Please run git submodule update --init --recursive.${DefaultColor}"
```
```suggestion
        message( FATAL_ERROR "${Red}The git submodule ${submodule} is not available. Please run git submodule update --init --recursive.${DefaultColor}"
```

vyasr replied:
The loop will terminate as soon as any one submodule fails, so specifying names will be misleading.

bdice replied:
But isn't it useful to know _which_ submodule raised the error? 🤨 The same solution applies regardless, but it seems more helpful than not identifying the submodule name.

vyasr replied:
I disagree, because almost nobody who's going to run into this issue in practice runs `git submodule update` on specific submodules. The most common error is not initializing submodules at all, and the people making that mistake are also the most likely to see this. Under those circumstances I think it's more confusing to see a single name, especially since the recommended solution doesn't specify a submodule. If you were to error on a specific submodule, the recommended solution should also be to update only that submodule.

bdice replied:
We disagree but I don't feel strongly, so you win. 😉

bdice commented on freud/CMakeLists.txt:
```diff
@@ -0,0 +1,117 @@
+    if (COVERAGE)
+        target_compile_definitions(
+            ${cython_module}
+            #  Enable line tracing for coverage purposes if requested.
```
```suggestion
            # Enable line tracing for coverage purposes if requested.
```

bdice commented on freud/CMakeLists.txt:
```diff
@@ -0,0 +1,117 @@
+    # Coverage requires the Cython-compiled C++ files for line coverage.
+    if (COVERAGE)
+        install(FILES ${${cython_module}} DESTINATION freud)
+	endif()
```
Tabs/spaces?
```suggestion
    endif()
```

bdice reviewed (commented):
We will also need to update our installation guide and developer documentation.

vyasr committed:
Move includes back to object libraries, only linking is not supported.

vyasr committed:
Force install of scikit-build prior to installing other requirements.

vyasr committed:
Some formatting improvements and minor improvements.

vyasr commented:
> We will also need to update our installation guide and developer documentation.

Yes, this is noted in the PR to-do list.

vyasr committed:
Merge remote-tracking branch 'origin/master' into feature/enable_skbuild

vyasr commented:
The Windows CI is currently failing because it appears that the available generators are not suitable for compiling the code. The scikit-build docs indicate that you should use the Visual Studio version used to build the Python distribution, but I'm not sure if it's actually that restrictive. @bdice I'd appreciate if you could give that a look, I have no experience with the Windows CI setup.

There is one significant (but IMO not disqualifying) regression currently associated with this PR. Due to some subtle issues with how these tools work together, Cython code coverage does not work with the new builds. I noticed the issue this morning and was able to identify the underlying issue by running the compilation manually and comparing Cython-generated C++ files, but I don't know how viable implementing a fix will be. I've posted a question to the Cython-users mailing list, we'll see what comes out of it.

@chr1shr indicated that he would be open to a PR improving the CMake configuration of voro++. In the interest of pushing progress forward on this PR, I would prefer to get this wrapped up and then submit a subsequent PR to change how we compile voro++. The current method is perfectly workable.

vyasr commented:
> Regarding `_util` I'm not sure what the problem is. As you say, maybe it is some limitation with object libraries. Since you are building everything into one `libfreud` target, you could instead make the child CMakeList.txt files set a list of source files in a variable and then have the `add_library` compile all those source files together. That would work around any limitations with `OBJECT` and wouldn't result in any files being compiled more than once.

Yes, this question has become less relevant since I originally created this PR due to the switch to compiling a single shared library for the C++ API. Additionally, the modifications I made to allow this compilation to work with older versions of CMake forced me to get rid of specifying libraries to link to Object libraries anyway, so most of the linking is now performed on the `libfreud` target anyway. 

> Regarding FindTBB, feel free to use the solution from HOOMD to support finding modern TBB's by their embedded config file (the `find_package_config_first` macro) - that way you don't need to rely on your own `FindTBB.cmake` except on systems like Ubuntu which doesn't provide the TBB config or on platforms with older TBB installations. The `FindTBB` from HOOMD determines the `TBB_VERSION` and implements a few compiler workarounds that may not be needed for freud. The only other main difference is that the target is `TBB::tbb` for compatibility with the TBB provided config files.

Sounds good, I'll take a look at that config and apply the relevant changes.

bdice committed:
Move Windows definitions to C++ CMakeLists.

vyasr committed:
Remove some minor redundancies in the CI script.

vyasr committed:
Switch to using FindTBB from HOOMD.

vyasr committed:
Merge branch 'feature/enable_skbuild' of https://github.com/glotzerlab/freud into feature/enable_skbuild

vyasr committed:
Temporarily enable linux PyPI deployment for testing.

vyasr committed:
Force installation of dependencies in deployment script, which ignores requirements.

vyasr committed:
Add dependencies of scikit-build.

vyasr committed:
Don't update homebrew, and allow dependency installation for build requirements.

vyasr committed:
Move to newer (but still relatively old) XCode version to continue testing sufficiently old but not deprecated versions.

vyasr committed:
Switch to newest XCode.

vyasr committed:
Add path on linux to ensure that cmake can be found.

vyasr committed:
Temporarily disable windows CI builds.

vyasr committed:
Print more information in deployment script.

vyasr committed:
Enable use of environment variables to specify TBB paths.

vyasr committed:
Print more information to debug OSX issues.

vyasr committed:
Temporarily disable builds to stop wasting CI hours during iterations on wheel building.

vyasr committed:
Manually update python before pyenv tries to do so automatically.

vyasr committed:
Only update 3.8...

vyasr committed:
Check if LD_LIBRARY_PATH exists before trying to use it. Stop running linux wheels for faster testing.

vyasr committed:
Remove extraneous colon.

vyasr committed:
Also respect DYLD_LIBRARY_PATH for Mac.

vyasr committed:
Actually check DYLD...

vyasr committed:
Don't test on MDAnalysis if it's not installed.

vyasr committed:
Update installation for wheels to skip MDAnalysis as well.

vyasr committed:
Clean up various intermediates states.

vyasr committed:
Remove fPIC again to see if I can make it work.

vyasr committed:
Add back missing colon.

vyasr committed:
Remove windows prereqs.

vyasr committed:
Looks like specifying PIC is indeed required and can't be inferred.

vyasr committed:
Add more comments.

vyasr committed:
Revert CI changes made for testing, outstanding task is to enable Windows support.

vyasr committed:
Update installation instructions.

vyasr committed:
Update the developer documentation on adding new code to freud.

bdice committed:
Merge remote-tracking branch 'origin/master' into feature/enable_skbuild

tommy-waltmann commented on cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+add_subdirectory(util)
+
+add_library(
+    libfreud
```
I believe this will create a library called liblibfreud.so, was the variable named knowing the library would be named like this?

bdice replied:
@tommy-waltmann This is intentional, it's for making it clear that this is the C++ library. The name is fixed later, so it ends up as "freud": https://github.com/glotzerlab/freud/blob/0618c150441e27a7dfe4cf4cb37174d9da188e45/cpp/CMakeLists.txt#L47

tommy-waltmann commented on cpp/cluster/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+    Cluster.h
+    Cluster.cc
+    ClusterProperties.h
+    ClusterProperties.cc
```
I don't think the headers are necessary here, or in other places like this.

bdice replied:
I didn't think so either, but I wasn't sure. @vyasr can you explain?

joaander replied:
Adding headers to a target is a convenience for IDE project targets in CMake - it makes the headers appear in the project. I don't know if it also factors into the compile_command.json output.

vyasr replied:
Yes exactly, some IDEs consume the CMakeLists.txt files directly. I'm not sure about compile_commands, but I think it's basically an issue of inconsistency between development environments where some will read the cmake file whereas others will look at JSON. I _think_ the compile_commands.json is necessary for semantic completion in certain contexts.

bdice committed:
Enable scikit-build on Windows (without static members) (#676)

Some minor modifications to the CMake scripts are necessary to enable building on Windows. Additionally, now that locality (and all of the C++ code) is being built once and linked to as a single shared library by all of the Cython modules, we need to make some minor modifications to the linkage of certain objects to avoid complications on Windows. In particular, static class members (like the ITERATOR_TERMINATOR) are now namespace-level constants to avoid the need for various MSVC-specific preprocessor directives.

vyasr committed:
Update RTD env.

vyasr committed:
Update changelog and credits.

vyasr committed:
Add notice that different generators might require newer generators.

vyasr commented:
I think this PR should be good to go now. I merged in the PR adding Windows support and augmented the docs a little bit. @bdice to address one of the issues that you ran into when enabling Windows support (requiring newer CMake versions) I added a note to our installation documentation indicating that this might be the case and pointing to the relevant CMake generator documentation.

bdice commented:
For some reason, the ReadTheDocs build is failing because it's uninstalling scikit-build.
https://readthedocs.org/projects/freud/builds/12216572/

```
$ conda install --yes --quiet --name 668 mock pillow sphinx sphinx_rtd_theme

[...] truncated [...]

The following packages will be REMOVED:

  jupyter_sphinx-0.3.1-py39hde42818_1
  pybtex-0.23.0-py39hde42818_0
  pybtex-docutils-0.2.2-py39hde42818_2
  python_abi-3.9-1_cp39
  scikit-build-0.11.1-py39hf149a3a_2
  sphinxcontrib-bibtex-1.0.0-py_0
```

bdice committed:
Swap channel priority for ReadTheDocs.

bdice commented:
I fixed the first problem with ReadTheDocs builds by swapping channel priority (`conda-forge` is now secondary). Now it's unable to find TBB. https://readthedocs.org/projects/freud/builds/12217822/
```
[42/79] Linking CXX shared library cpp/libfreud.so
FAILED: cpp/libfreud.so 
: && /usr/bin/c++ -fPIC -O3 -DNDEBUG   -shared -Wl,-soname,libfreud.so -o cpp/libfreud.so cpp/cluster/CMakeFiles/_cluster.dir/Cluster.cc.o cpp/cluster/CMakeFiles/_cluster.dir/ClusterProperties.cc.o cpp/density/CMakeFiles/_density.dir/CorrelationFunction.cc.o cpp/density/CMakeFiles/_density.dir/GaussianDensity.cc.o cpp/density/CMakeFiles/_density.dir/LocalDensity.cc.o cpp/density/CMakeFiles/_density.dir/RDF.cc.o cpp/density/CMakeFiles/_density.dir/SphereVoxelization.cc.o cpp/environment/CMakeFiles/_environment.dir/AngularSeparation.cc.o cpp/environment/CMakeFiles/_environment.dir/BondOrder.cc.o cpp/environment/CMakeFiles/_environment.dir/LocalBondProjection.cc.o cpp/environment/CMakeFiles/_environment.dir/LocalDescriptors.cc.o cpp/environment/CMakeFiles/_environment.dir/MatchEnv.cc.o cpp/locality/CMakeFiles/_locality.dir/AABBQuery.cc.o cpp/locality/CMakeFiles/_locality.dir/LinkCell.cc.o cpp/locality/CMakeFiles/_locality.dir/NeighborComputeFunctional.cc.o cpp/locality/CMakeFiles/_locality.dir/NeighborList.cc.o cpp/locality/CMakeFiles/_locality.dir/PeriodicBuffer.cc.o cpp/locality/CMakeFiles/_locality.dir/Voronoi.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/cell.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/common.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/container.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/unitcell.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/v_compute.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/c_loops.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/v_base.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/wall.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/pre_container.cc.o cpp/locality/CMakeFiles/_locality.dir/__/__/extern/voro++/src/container_prd.cc.o cpp/order/CMakeFiles/_order.dir/Cubatic.cc.o cpp/order/CMakeFiles/_order.dir/HexaticTranslational.cc.o cpp/order/CMakeFiles/_order.dir/Nematic.cc.o cpp/order/CMakeFiles/_order.dir/RotationalAutocorrelation.cc.o cpp/order/CMakeFiles/_order.dir/SolidLiquid.cc.o cpp/order/CMakeFiles/_order.dir/Steinhardt.cc.o cpp/order/CMakeFiles/_order.dir/Wigner3j.cc.o cpp/parallel/CMakeFiles/_parallel.dir/tbb_config.cc.o cpp/pmft/CMakeFiles/_pmft.dir/PMFTR12.cc.o cpp/pmft/CMakeFiles/_pmft.dir/PMFTXY.cc.o cpp/pmft/CMakeFiles/_pmft.dir/PMFTXYT.cc.o cpp/pmft/CMakeFiles/_pmft.dir/PMFTXYZ.cc.o cpp/util/CMakeFiles/_util.dir/diagonalize.cc.o  -Wl,-rpath,/home/docs/checkouts/readthedocs.org/user_builds/freud/conda/668/lib:  /home/docs/checkouts/readthedocs.org/user_builds/freud/conda/668/lib/libtbb.so && :
/usr/bin/ld: cannot find libtbb.so.2
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
  File "/home/docs/checkouts/readthedocs.org/user_builds/freud/conda/668/lib/python3.8/site-packages/skbuild/setuptools_wrap.py", line 589, in setup
    cmkr.make(make_args, env=env)
  File "/home/docs/checkouts/readthedocs.org/user_builds/freud/conda/668/lib/python3.8/site-packages/skbuild/cmaker.py", line 496, in make
    raise SKBuildError(
```

vyasr commented:
I don't know offhand what's going on there. It _is_ finding TBB, CMake prints the location and it's being provided as an additional target for the compiler on that invocation of `c++`. What it's not finding is `libtbb.so.2`, which I'm assuming is an artifact of this particular build of TBB provided by conda where `libtbb.so` needs `libtbb.so.2`. It seems like the way that RTD uses the conda environment is to manually provide all the absolute paths rather than to activate the environment (based on looking at the build link you pasted I never see a `conda activate`) so things that aren't explicitly added aren't found. I assume we don't have any control at that level of the build?

Actually I take that back. It looks like the conda library directory is being added to the rpath. Then I'm not sure why it wouldn't find this file, unless it's actually not present.

tommy-waltmann commented on tests/integration/test_reader_integrations.py:
```diff
@@ -5,6 +5,12 @@
 import sys
 import unittest
 
+try:
+    import MDAnalysis
+    MDANALYSIS = True
+except ImportError:
+    MDANALYSIS = False
+
 
```
Why is this, and the other logic changes in this file, a part of this PR?

vyasr replied:
MDAnalysis does not provide wheels, and it's not compatible with the newest clang versions, so compiling will fail outright. This causes problems for our wheel builds because we run tests on CI that depend on MDAnalysis (the integration tests). The newest clang releases happened after our last release of freud, and we only build with clang when we build OS X wheels at release, so we simply hadn't run into this problem yet. Since I identified the issues when modifying the wheel building process to work with scikit-build, I had to make these changes to test the whole thing. Ideally it wouldn't be here, but even if I did it in a separate PR it would immediately need to be merged here, so I just left it as part of this PR.

bdice commented on ChangeLog.md:
```diff
@@ -12,6 +12,7 @@ and this project adheres to
 
 ### Changed
 * NeighborList raises a `ValueError` instead of a `RuntimeError` if provided invalid constructor arguments.
+* freud now builds using scikit-build.
```
```suggestion
* freud now builds using scikit-build (requires CMake).
```

bdice commented on doc/source/gettingstarted/installation.rst:
```diff
@@ -64,47 +81,37 @@ You can also build **freud** in place so that you can run from within the folder
+CMake Options
++++++++++++++
+
+The scikit-built tool allows setup.py to accept three different sets of options separated by ``--``, where each set is provided directly to scikit-build, to CMake, or to the code generator of choice, respectively.
```
```suggestion
The scikit-build tool allows setup.py to accept three different sets of options separated by ``--``, where each set is provided directly to scikit-build, to CMake, or to the code generator of choice, respectively.
```

bdice commented on doc/source/gettingstarted/installation.rst:
```diff
@@ -64,47 +81,37 @@ You can also build **freud** in place so that you can run from within the folder
 
-    -\\-PRINT-WARNINGS
-      Specify whether or not to print compilation warnings resulting from the build even if the build succeeds with no errors.
+    --COVERAGE
```
This may need a `\\`, I noticed this pattern below.
```suggestion
    -\\-COVERAGE
```

vyasr replied:
Actually that pattern is wrong, I'm not sure why it got introduced. It renders a backslash in between the two dashes, [see the docs right now](https://freud.readthedocs.io/en/stable/gettingstarted/installation.

bdice replied:
I fixed it.

bdice replied:
For posterity, this is the escaping behavior for smart double-dashes in docutils. https://docutils.sourceforge.io/docs/user/smartquotes.html#id5

bdice committed:
Update credits.

bdice commented on doc/source/reference/development/howtoadd.rst:
```diff
@@ -54,7 +54,7 @@ To keep modules well-organized, **freud** implements the following structure:
 - C++ code is exposed to Python using Cython code contained in pxd files with the following convention: ``freud/_MODULENAME.pxd`` (note the preceding underscore).
 - The core Cython code for modules is contained in ``freud/MODULENAME.pyx`` (no underscore).
-- Generated Cython C++ code (e.g. ``freud/MODULENAME.cpp``) should not be committed during development. These files are updated only at release time.
+- Generated Cython C++ code (e.g. ``freud/MODULENAME.cpp``) should not be committed during development. These files are generated using Cython when building from source, and are unnecessary when installing compiled binaries.
```
Are these actually named `*.cxx` now? I noticed that in the `.gitignore`.
```suggestion
- Generated Cython C++ code (e.g. ``freud/MODULENAME.cxx``) should not be committed during development. These files are generated using Cython when building from source, and are unnecessary when installing compiled binaries.
```

vyasr replied:
Yep that's the default output now.

bdice replied:
@vyasr Did you mean to accept my suggestion? You seemed to agree with my comment, which I think would indicate that we need to change `cpp` to `cxx` in this line about generated code, but you resolved the thread without applying the suggestion.

vyasr replied:
Yes, I must have missed clicking it when I went through to batch the suggestions.

bdice commented on doc/source/reference/development/howtoadd.rst:
```diff
@@ -84,8 +84,13 @@ Steps for Adding New Code
+If the new methods belong in a new module, you must create the corresponding ``cpp`` directory and the ``pxd`` and ``pyx`` files accordingly.
+
+In order for code to compile, it must be added to the relevant ``CMakeLists.txt`` file.
+New C++ files for existing modules must be added to the corresponding ``cpp/MODULENAME/CmakeLists.txt`` file.
```
```suggestion
New C++ files for existing modules must be added to the corresponding ``cpp/MODULENAME/CMakeLists.txt`` file.
```

bdice commented on doc/source/reference/development/howtoadd.rst:
```diff
@@ -84,8 +84,13 @@ Steps for Adding New Code
+
+In order for code to compile, it must be added to the relevant ``CMakeLists.txt`` file.
+New C++ files for existing modules must be added to the corresponding ``cpp/MODULENAME/CmakeLists.txt`` file.
+For new modules, a ``cpp/NEWMODULENAME/CMakeLists.txt`` file must be created, and in addition the new module must be added to the ``cpp/CMakeLists`` file in the form of both an ``add_subdirectory`` command and addition to the ``libfreud`` library in the form of an additional source in the ``add_library`` command.
```
```suggestion
For new modules, a ``cpp/NEWMODULENAME/CMakeLists.txt`` file must be created, and in addition the new module must be added to the ``cpp/CMakeLists.txt`` file in the form of both an ``add_subdirectory`` command and addition to the ``libfreud`` library in the form of an additional source in the ``add_library`` command.
```

vyasr commented:
OK I'm very confused now. As far as I can tell the tbb libraries on conda-forge right now are missing libtbb.so.2, but libtbb.so always points to it. I'm not sure how this works at all, but I'm triggering a rebuild of master on RTD to make sure this isn't some recent change in the tbb libraries. It doesn't seem like it (looking backward to older tbb versions), but I would like to verify.

bdice commented on freud/CMakeLists.txt:
```diff
@@ -0,0 +1,108 @@
+# including the header requires the Cluster.h header. Would prefer to inherit
+# this information from the _order library, but that's not possible since we're
+# linking to libfreud.
+target_include_directories(
```
Is there a way to make each module include the "include directories" in `${PROJECT_SOURCE_DIR}/cpp/${cython_module}/CMakeLists.txt`?

vyasr replied:
Since we're linking against libfreud, I think any (not super convoluted) attempt to use the CMakeLists files in the cpp folders will lead to multiple builds of the same objects and linker errors. I don't think you can selectively include parts of another CMakeLists.txt file.

bdice replied:
That sounds fine to me, then. Thanks for the answer.

bdice reviewed (changes requested):
Looks good overall! We just need to fix a few minor issues and make ReadTheDocs work.

vyasr commented:
Ah OK I found it. The real library is installed by the `tbb` package on conda-forge. My guess is that the old Cython install was finding all of this, but the new install finds `libtbb.so` and thinks it's good enough. I'll need to improve the TBB finding so that it uses the config when available, that probably will fix it. I'll try to get around to this later today.

vyasr committed:
Use CMake config to find packages if available.

vyasr committed:
Use consistent set of variables irrespective of whether config is used for TBB or not.

vyasr committed:
Try switching Windows TBB linking back.

vyasr committed:
Only use the special Windows behavior in the one case we know it's absolutely necessary.

vyasr committed:
Apply suggestions from code review

Co-authored-by: Bradley Dice <bdice@bradleydice.com>

vyasr committed:
Merge branch 'master' into feature/enable_skbuild

bdice committed:
Lint whitespace in CMakeLists.txt.

vyasr commented:
Actually hold off on calling this completely done. I had intended to format the cmake files using a standard linter and add that to CI. I had found something decent and will apply it before merging this pr.

bdice committed:
Fix escaped dashes.

tommy-waltmann commented on CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/CMake")
+include(${CMAKE_CURRENT_LIST_DIR}/CMake/freud-macros.cmake)
```
I think it would be more modular to call `add_subdirectory(CMake)` here, and then in that directory's CMakeLists have these two lines.

vyasr replied:
Good call, I'll make that change.

bdice committed:
Revert "Swap channel priority for ReadTheDocs."

This reverts commit 9ec9f2432d858535a4e6a1eb81e150636334104f.

vyasr committed:
Format using cmake-format.

vyasr committed:
Create CMakeLists.txt file for CMake directory.

vyasr committed:
Merge branch 'feature/enable_skbuild' of https://github.com/glotzerlab/freud into feature/enable_skbuild

vyasr committed:
Update doc/source/reference/development/howtoadd.rst

Co-authored-by: Bradley Dice <bdice@bradleydice.com>

vyasr commented:
I applied a reformatting. I'll make a separate PR to make this part of CI etc so that we can get this merged and in production.

vyasr committed:
Fix path to FindTBB.

vyasr committed:
Merge branch 'feature/enable_skbuild' of https://github.com/glotzerlab/freud into feature/enable_skbuild

vyasr committed:
Set scope for the modification of the path.

vyasr commented:
RTD builds are back to failing, but now it's because the build has upgraded to using Python 3.9, scipy wheels aren't available yet, and compiling scipy fails. I think we can probably just go ahead and merge this PR at this point? This problem won't be resolved until scipy 3.9 wheels are available.

vyasr committed:
Fix double hyphens in rst.

bdice reviewed (approved):
Looks good to me. Merge when you're ready. Thanks for the scikit- _boooo_ ild
![spooky parrot](https://cultofthepartyparrot.com/parrots/hd/pumpkinparrot.gif)

vyasr committed:
Document environment variables appropriately.

vyasr commented:
@tommy-waltmann in the interest of keeping things moving along, I went ahead and merged. If you have further comments though, please do open another PR (or just an issue) with proposed changes!
